### PR TITLE
Add gql_websocket_link adapting code from graphql-flutter

### PR DESCRIPTION
**Pull request by iscriptology**
This is an adapted version of the web socket link in the `graphql-flutter` repo.
There are some minor changes I made:
1. Prevent the web socket from connecting until first request (as stated in the docs)
2. Making the code `pedantic`
3. Adapting it to the `gql_link` interface.
4. Copied the relevant tests files as well and made them pass. 

Additionally I manually tested it and it seems to work just fine.
@comigor @klavs @smkhalsa code review and merge would be great as it is necessary for GraphQL subscriptions. (`ferry` lacks them for example, due to the absence of this link)

### CODE REVIEW

**Comment by iscriptology** [comment]
#76

**Comment by klavs** [comment]
Thank you @iscriptology!
Can you coordinate with @agent3bood to make sure both of your efforts are heading the same direction? see https://github.com/gql-dart/gql/pull/93

File: gql_ws_link/lib/src/socket_client.dart [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+
+        final Stream<GraphQLSocketMessage> subscriptionComplete = addTimeout
+            ? dataErrorComplete
+            .where((GraphQLSocketMessage message) =>
+        message is SubscriptionComplete)
+            .take(1)
+            .timeout(
+          config.queryAndMutationTimeout,
+          onTimeout: (EventSink<GraphQLSocketMessage> event) {
+            print("Request timed out.");
```
Comment (klavs) [inline gql_ws_link/lib/src/socket_client.dart:311]: The `print` statements should be dropped. People who want to debug should opt-in using the debugger.

File: gql_ws_link/test/gql_ws_link_test.dart [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+        ),
+        variables: const <String, dynamic>{"i": 12},
+      );
+      link = WSLink(
+        "/graphql-test",
+        socketClient: client,
+      );
+    });
+
+//    test("parses a successful response", () {
```
Comment (klavs) [inline gql_ws_link/test/gql_ws_link_test.dart:51]: I'd rather see these removed or replaced with a link to the original.

File: gql_ws_link/test/helpers.dart [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+dynamic Function() overridePrint(Function(List<String> log) testFn) => () {
+  final log = <String>[];
+  final spec = ZoneSpecification(print: (_, __, ___, String msg) {
+    log.add(msg);
+  });
+  return Zone.current.fork(specification: spec).run<dynamic>(() => testFn(log));
+};
+
+//NormalizedInMemoryCache getTestCache() => NormalizedInMemoryCache(
```
Comment (klavs) [inline gql_ws_link/test/helpers.dart:20]: Can be removed.

File: gql_ws_link/lib/src/link.dart [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import "package:gql_ws_link/src/messages.dart";
+import "package:gql/language.dart";
+import "package:gql/ast.dart";
+
+/// A Universal WebSocket [Link] implementation to support the WebSocket-GraphQL transport.
+/// It supports subscriptions, query and mutation operations as well.
+///
+/// NOTE: the actual socket connection will only get established after a [Request] is handled by this [WSLink].
+/// If you'd like to connect to the socket server instantly, call the [connectOrReconnect] method after creating this [WSLink] instance.
+class WSLink extends Link {
```
Comment (klavs) [inline gql_ws_link/lib/src/link.dart:16]: What do you think about renaming this to `WebSocketLink`?

File: gql_ws_link/pubspec.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+name: gql_ws_link
```
Comment (klavs) [inline gql_ws_link/pubspec.yaml:1]: What do you think about renaming this to `gql_websocket_link`?

File: gql_ws_link/pubspec.yaml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+version: 0.1.0
+homepage: https://github.com/gql-dart/gql/tree/master/gql_ws_link
+description: GQL Websocket Link
+environment: 
+  sdk: '>=2.7.2 <3.0.0'
+dependencies: 
+  meta: ^1.1.8
+  gql_exec: ^0.2.4
```
Comment (comigor) [inline gql_ws_link/pubspec.yaml:9]: It's best to require all `gql*` packages by relative path

File: gql_ws_link/lib/src/link.dart [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  final String url;
+  SocketClient _client;
+  final SocketClientConfig config;
+
+  ResponseParser parser;
+
+  WSLink(this.url, {this.config, this.parser = const ResponseParser()});
```
Comment (comigor) [inline gql_ws_link/lib/src/link.dart:23]: It's best to make `WSLink` to receive custom headers (and certificates if possible), as some servers require custom authentication, just like HTTPLink does: https://github.com/gql-dart/gql/blob/master/gql_http_link/lib/src/link.dart#L73-L74

Comment (iscriptology) [reply gql_ws_link/lib/src/link.dart:23]: There’s connectionParams for that.
Problem with headers, is that browser-environments don’t allow customizing then when performing the WebSocket HTTP upgrade request

File: gql_ws_link/pubspec.yaml (hunk 6 above)
Comment (klavs) [reply gql_ws_link/pubspec.yaml:9]: I think that's not true. Pub does not allow to publish with relative deps.

**Comment by iscriptology** [comment]
seems like dartfmt fails the checks... Executed it locally but not changing anything... Anyone?

**Comment by klavs** [comment]
@iscriptology what's your dartfmt version? `dartfmt --version`

File: gql_ws_link/pubspec.yaml (hunk 6 above)
Comment (comigor) [reply gql_ws_link/pubspec.yaml]: oh right, my bad!

File: gql_ws_link/lib/src/link.dart (hunk 7 above)
Comment (comigor) [reply gql_ws_link/lib/src/link.dart]: hmm, I'm not familiar with websockets, but is there some kind of pattern on authentication then? (which works on the browser)

**Comment by iscriptology** [comment]
`1.3.2`

**Comment by klavs** [comment]
`dartfmt -w .` from `gql/gql_websocket_link` worked for me. (I'm on `1.3.6`)
```
Formatting directory .:
Unchanged example/gql_websocket_link_example.dart
Unchanged lib/gql_websocket_link.dart
Unchanged lib/src/exceptions.dart
Formatted lib/src/get_from_ast.dart
Formatted lib/src/link.dart
Formatted lib/src/messages.dart
Formatted lib/src/socket_client.dart
Unchanged test/gql_websocket_link_test.dart
Formatted test/helpers.dart
Formatted test/socket_client_test.dart

```

**Comment by iscriptology** [comment]
I would suggest publishing and fix issues as they are reported. No web-socket link is a major drawback to the entire ecosystem of dependant packages.
Also to get in touch with the fellows at `graphql_flutter` to try and utilise the shared implementation and contribute to it.

**Comment by klavs** [comment]
Thank you @iscriptology! This is a key piece for making `gql_link` a viable replacement for `graphql_flutter` links.

I'll try to publish it in the next 24h.

**Comment by klavs** [comment]
I was lucky. My automated alpha version publishing workflow worked and this is now published as https://pub.dev/packages/gql_websocket_link

File: gql_ws_link/lib/src/link.dart (hunk 7 above)
Comment (iscriptology) [reply gql_ws_link/lib/src/link.dart]: You can send custom payload upon the socket creation...
In Apollo's case they used something called "connectionParams" which is far less standardised than HTTP headers but it's fine

**Comment by iscriptology** [comment]
Looking forward to @agent3bood contribution and to see this implementation matures

**Pull request merged**